plutus-core/test/Evaluation/Golden.hs

```diff
@@ -105,26 +114,172 @@ closure :: uni `Includes` Integer => Term TyName Name uni fun ()
 closure = runQuote $ do
     i <- freshName "i"
     j <- freshName "j"
-    let integer = mkTyBuiltin @Integer ()
+    let integer' = mkTyBuiltin @Integer ()
     pure
-        . Apply () (LamAbs () i integer . LamAbs () j integer $ Var () i)
+        . Apply () (LamAbs () i integer' . LamAbs () j integer' $ Var () i)
         $ mkConstant @Integer () 1
 
+{- Tests for evaluation of builtins, mainly checking that interleaving of term
+   and type arguments is correct.  At the moment the only polymorphic builtin
+   we have is ifThenElse. -}
+
+-- Various components that we'll use to build larger terms for testing
+
+lte :: Term TyName Name DefaultUni DefaultFun ()
+lte = Builtin () LessThanEqInteger
+
+eleven :: Term TyName Name DefaultUni DefaultFun ()
+eleven = mkConstant @Integer () 11
+
+twentytwo :: Term TyName Name DefaultUni DefaultFun ()
+twentytwo = mkConstant @Integer () 22
+
+stringResultTrue :: Term TyName Name DefaultUni DefaultFun ()
+stringResultTrue = mkConstant @String () "11 <= 22"
+
+stringResultFalse :: Term TyName Name DefaultUni DefaultFun ()
+stringResultFalse = mkConstant @String () "¬(11 <= 22)"
+
+-- 11 <= 22
+lteExpr :: Term TyName Name DefaultUni DefaultFun ()
+lteExpr = mkIterApp () lte [eleven, twentytwo]
+
+
+-- Various combinations of (partial) instantiation/application for ifThenElse
+
+-- (builtin ifThenElse)
+ite :: Term TyName Name DefaultUni DefaultFun ()
+ite = Builtin () IfThenElse
+
+-- { (builtin ifThenElse) (con integer) }
+iteAtInteger :: Term TyName Name DefaultUni DefaultFun ()
+iteAtInteger = TyInst () ite integer
+
+-- { (builtin ifThenElse) (con string) }
+iteAtString :: Term TyName Name DefaultUni DefaultFun ()
+iteAtString = TyInst () ite string
+
+-- { (builtin ifThenElse) (forall a . a -> a) }
+-- Evaluation should succeed, but typechecking should fail.
+-- You're not allowed to instantiate a builtin at a higher kind.
```

Comment:
Wait... why? That would be a huge restriction. And if it's the case, then it's a bug. I don't think it's the case, but maybe I'm missing something?

Reply:
Is that not true?  I'm sure we had that restriction at some point (maybe with the earlier version of the CEK machine).  You're definitely only allowed to instantiate `ifThenElse` at kind `*` (which you'll know, since you implemented it):
```
$ plc evaluate --typed -i ifhk.plc
[ { (builtin ifThenElse) (lam A (type) (fun A A)) } (con bool True) ]

$ plc typecheck -i ifhk.plc
Kind mismatch at () in type '(lam A (type) (fun A A))'. Expected kind '(type)' , found kind '(fun (type) (type))'
```
I thought that restriction was universal for some reason. If not, I'll fix the comment.

Also: this is why we need a specification!

Reply:
> You're definitely only allowed to instantiate ifThenElse at kind * (which you'll know, since you implemented it)

Sure, but that has nothing to do with builtins: whatever function expects a type argument of kind `*`, has to receive a type argument of kind `*`. It's not a restriction, but rather just regular typing.

I think we have a confusion here. The test case and the comment above it talk about `forall a. a -> a`, but your latest comment is about `lam a. a -> a`. It's sure the case that you can't instantiate what expects a `*` with a `* -> *`, but `forall a. a -> a` is a `*`. I.e. it's not a higher kind (contrary to what the comment above the test says), but rather a higher-rank type. Which should be perfectly fine (contrary to what the same comment says. Unless I misunderstand something terribly).

Reply:
Ah! I wrote the wrong thing in the test.  It was supposed to be what I wrote above, but I may have been asleep when I actually wrote the test.  Thanks for pointing that out: I'll fix it, and maybe also add a test that higher order _types_ are OK.

Regarding the kinds, is it in fact the case that builtins can only be instantiated at `*`, or are we allowed to implement a builtin that's polymorphic over something of kind `* -> *`, for example?  I don't know if we'd ever _want_ to do that, but does the builtin machinery permit it?  I definitely had the idea that that wasn't allowed, but I don't remember why (and maybe things have changed since then).

Reply:
>  Regarding the kinds, is it in fact the case that builtins can only be instantiated at `*`, or are we allowed to implement a builtin that's polymorphic over something of kind `* -> *`, for example? I don't know if we'd ever want to do that, but does the builtin machinery permit it?

It does, I implemented that some (not long) time ago. I don't know if it's useful for anything, but it was fairly trivial to implement, so why not. See [this test](https://github.com/input-output-hk/plutus/blob/397ae9600b385b3239b1f72bbcad72e74d517107/plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs#L230).